src/style-installer.js

```diff
@@ -283,6 +270,25 @@ export function makeBodyVisible(doc, opt_waitForServices) {
   }
 }
 
+/**
+ * Set the document's body opacity to 1. Called in error cases.
+ * @param {!Document} doc The document who's body we should make visible.
+ */
+export function makeBodyVisibleRecovery(doc) {
+  dev().assert(doc.defaultView, 'Passed in document must have a defaultView');
+  const win = /** @type {!Window} */ (doc.defaultView);
+  if (win[bodyVisibleSentinel]) {
+    return;
+  }
+
+  win[bodyVisibleSentinel] = true;
```

Comment:
If the issue is we can't trust caching on `window`, why not use an attribute on the body instead? We could remove this `setStyles` and move the actual CSS into `amp.css`.

Reply:
Given the nature of the Chrome bug, I believe `set attribute` would not work around it.

But +1 to use CSS in amp.css, so we don't duplicate code here.

Reply:
> Given the nature of the Chrome bug, I believe set attribute would not work around it.

The reason I propose it is because using an attribute means we won't need to do `win[bodyVisibleSentinel]` check at all.

Actually, looking at https://github.com/ampproject/amphtml/issues/4822, we really only need to guard against running `renderStartedNoInline` twice. Everything else doesn't really matter, unless somewhere else we're applying some inline styles to body. An attribute won't have an overriding affect on that, hence we can just `setAttribute` without any checks.

Reply:
This is slightly misleading naming, but the big reason is also avoiding setting any styles on the body in shadow mode where we don't want to touch `body` at all.

Also, I think we can switch from `win[bodyVisibleSentinel]` to `let bodyMadeVisible;` module var.

Reply:
@dvoytenko is right. CSS selector on body will break PWA case.

> Also, I think we can switch from win[bodyVisibleSentinel] to let bodyMadeVisible; module var.

I guess it's safe for now, since all the use cases right now having accessing the var in the same script binary. Let's document it clearly on this method `bodyAlwaysVisible`.

We can think of renaming in a separate PR.

Reply:
Moved to module level var.